### PR TITLE
ZDC - Fix raw data checks, better debug output, speedup reconstruction

### PR DESCRIPTION
**Pull request by cortesep**
Found bug in checks performed on ALICE triggers (..were not present in simulation). Improved debug output in case of inconsistent raw data.

### CODE REVIEW

File: Detectors/ZDC/raw/src/DumpRaw.cxx [hunk 1]
```diff
@@ -56,7 +57,57 @@ void DumpRaw::init()
...
+  double xmax = 2 * NTimeBinsPerBC - 0.5;
+  if (mTransmitted == nullptr) {
+    mTransmitted = (TH2*)gROOT->FindObject("ht");
+  }
+  if (mTransmitted == nullptr) {
+    mTransmitted = new TH2F("ht", "Transmitted channels", NModules, -0.5, NModules - 0.5, NChPerModule, -0.5, NChPerModule - 0.5);
+  }
+  if (mBits == nullptr) {
+    mBits = (TH2*)gROOT->FindObject("hb");
+  }
```
Comment (shahor02) [inline Detectors/ZDC/raw/src/DumpRaw.cxx]: @cortesep could you explain where the raw-parser using this class is running and where goes this histo file?
In any case, what is the reason of creating a "stateless" device https://github.com/AliceO2Group/AliceO2/blob/226842d256ae68032c9a82fe1b0fc07511246a6c/Detectors/ZDC/raw/src/raw-parser.cxx#L55 which in fact needs to book a bunch of histos and update them from TF to TF? Also, writing the updated histos to a file at every TF is an overkill. Is this production workflow?

Comment (cortesep) [reply Detectors/ZDC/raw/src/DumpRaw.cxx]: @shahor02 This is not a production workflow. It is used on raw data files that are produced at P2 and copied outside the ALICE network. Usually the histograms I develop into Detectors/ZDC/raw/include/ZDCRaw/RawReaderZDC.h are later introduced in the QC worflow (that has, by the way, an excessively long upload latency).

For what concerns the executable raw-parser I just copied an example some time ago and I don't understand well what it is doing. It seems it is launched at every time frame and therefore I have to call gROOT->FindObject at every step in order to continue filling the histograms. Do you have a suggestion how can I improve this? Thanks.

Comment (shahor02) [reply Detectors/ZDC/raw/src/DumpRaw.cxx]: > For what concerns the executable raw-parser I just copied an example some time ago and I don't understand well what it is doing. It seems it is launched at every time frame and therefore I have to call gROOT->FindObject at every step in order to continue filling the histograms. Do you have a suggestion how can I improve this? Thanks.

well, the suggestion would be to convert it to usual DPL task which defines histos at init then fills them in the run(..).
But, if this is only for private debugging, I don't really care about its efficiency :)

Comment (cortesep) [reply Detectors/ZDC/raw/src/DumpRaw.cxx]: Ok. I will try to modify this.

Comment (cortesep) [reply Detectors/ZDC/raw/src/DumpRaw.cxx]: Thanks for the hint. Speed improvement is huge. Now 1 GB raw data file is analyzed in just one minute.

**Comment by cortesep** [comment]
Hi Ruben,

do you think it is ok now?
Thanks,
Pietro

File: Detectors/ZDC/raw/include/ZDCRaw/DumpRaw.h [hunk 2]
```diff
@@ -39,13 +40,17 @@ class DumpRaw
  private:
   void setStat(TH1* h);
   int mVerbosity = 1;
+  TH2* mTransmitted = nullptr;
```
Comment (shahor02) [inline Detectors/ZDC/raw/include/ZDCRaw/DumpRaw.h]: Hi @cortesep 
Please, either use unique_ptr for pointers (preferred) or add a destructor where you cleanup the histos

**Review by shahor02 — CHANGES_REQUESTED** [review]

File: Detectors/ZDC/raw/include/ZDCRaw/DumpRaw.h (hunk 2 above)
Comment (cortesep) [reply Detectors/ZDC/raw/include/ZDCRaw/DumpRaw.h]: Hi @shahor02,
Introduced unique_ptr for the histograms

**Review by shahor02 — APPROVED** [review]

**Pull request merged**